theano/gpuarray/opt.py

```diff
@@ -165,25 +169,36 @@ def f(local_opt):
 
 def safe_to_gpu(x, ctx_name):
     if isinstance(x.type, tensor.TensorType):
-        return gpu_from_host(ctx_name)(x)
+        return GpuFromHost(ctx_name)(x)
     else:
         return x
 
 
 def safe_to_cpu(x):
     if isinstance(x.type, GpuArrayType):
-        return host_from_gpu(x)
+        return x.transfer('cpu')
     else:
         return x
 
 
+def register_topo(*tags, **kwargs):
+    def f(local_opt):
+        name = (kwargs and kwargs.pop('name')) or local_opt.__name__
+        name = name + "_graph"
+        gpu_topo.register(name, local_opt,
+                          10, 'fast_run', 'gpuarray', *tags)
+        return local_opt
+    return f
```

Comment:
During July, @lamblin proposed this, "I'm proposing to have a second pass in the other direction: from outputs towards the inputs, where we could lift the remaining Ops by their output.
Actually, both passes could also become part of the GraphToGPUOptimizer. This remain to be decided. "
Now it makes sense for me why I registered a topo optimizer. It doesn't give any change (or shouldn't) in profiling immediately. But when we implement backward pass, there should be. But earlier we stopped from implementing a backward pass, not knowing the exact details of number of nodes that weren't moved to the GPU in the primary pass and whether it is worth implementing one.